docs/src/neural_ode/neural_gde.md

```diff
@@ -1,146 +1,263 @@
+using MLDatasets: Cora
+using CUDA
+CUDA.allowscalar(false)
+device = CUDA.functional() ? gpu : cpu
 
 # Download the dataset
-download("https://rawcdn.githack.com/yuehhua/GeometricFlux.jl/a94ca7ce2ad01a12b23d68eb6cd991ee08569303/data/cora_features.jld2", "cora_features.jld2")
-download("https://rawcdn.githack.com/yuehhua/GeometricFlux.jl/a94ca7ce2ad01a12b23d68eb6cd991ee08569303/data/cora_graph.jld2", "cora_graph.jld2")
-download("https://rawcdn.githack.com/yuehhua/GeometricFlux.jl/a94ca7ce2ad01a12b23d68eb6cd991ee08569303/data/cora_labels.jld2", "cora_labels.jld2")
+dataset = Cora()
+
+# Preprocess the data and compute adjacency matrix
+classes = dataset.metadata["classes"]
+g = mldataset2gnngraph(dataset) |> device
+onehotbatch(data,labels)= device(labels).==reshape(data, 1,size(data)...)
+onecold(y) =  map(argmax,eachcol(y))
+X = g.ndata.features
+y = onehotbatch(g.ndata.targets, classes) # a dense matrix is not the optimal, but we don't want to use Flux here
 
-# Load the dataset
-@load "./cora_features.jld2" features
-@load "./cora_labels.jld2" labels
-@load "./cora_graph.jld2" g
+Ã = normalized_adjacency(g, add_self_loops=true) |> device
+
+(; train_mask, val_mask, test_mask) = g.ndata
+ytrain = y[:,train_mask]
 
 # Model and Data Configuration
-num_nodes = 2708
-num_features = 1433
-hidden = 16
-target_catg = 7
-epochs = 40
+nin = size(X, 1)
+nhidden = 16
+nout = length(classes)
+epochs = 20
+
+# Define the graph neural network
+struct ExplicitGCNConv{F1,F2,F3,F4} <: AbstractExplicitLayer
+    in_chs::Int
+    out_chs::Int
+    activation::F1
+    init_Ã::F2  # nomalized_adjacency matrix
+    init_weight::F3
+    init_bias::F4
+end
 
-# Preprocess the data and compute adjacency matrix
-train_X = Float32.(features)  # dim: num_features * num_nodes
-train_y = Float32.(labels)  # dim: target_catg * num_nodes
+function Base.show(io::IO, l::ExplicitGCNConv)
+    print(io, "ExplicitGCNConv($(l.in_chs) => $(l.out_chs)")
+    (l.activation == identity) || print(io, ", ", l.activation)
+    print(io, ")")
+end
+
+function initialparameters(rng::AbstractRNG, d::ExplicitGCNConv)
+        return (weight=d.init_weight(rng, d.out_chs, d.in_chs),
+                bias=d.init_bias(rng, d.out_chs, 1))
+end
 
-adj_mat = FeaturedGraph(Matrix{Float32}(adjacency_matrix(g)))
+initialstates(rng::AbstractRNG, d::ExplicitGCNConv) = (Ã = d.init_Ã(),)
+
+
+function ExplicitGCNConv(Ã, ch::Pair{Int,Int}, activation = identity;
+                         init_weight=glorot_normal, init_bias=zeros32) 
+    init_Ã = ()->copy(Ã)
+    return ExplicitGCNConv{typeof(activation), typeof(init_Ã), typeof(init_weight), typeof(init_bias)}(first(ch), last(ch), activation, 
+                                                                                                       init_Ã, init_weight, init_bias)
+end
+
+function (l::ExplicitGCNConv)(x::AbstractMatrix, ps, st::NamedTuple)
+    z = ps.weight * x * st.Ã
+    return l.activation.(z .+ ps.bias), st
+end
 
```

Comment:
It would be nice if we have a GNN framework based on Lux. @CarloLucibello any comments?

Reply:
I wouldn't mind adding primitive GNNs to Lux, though it needs tests and docs before it can be added.

Reply:
the whole message passing api in https://github.com/CarloLucibello/GraphNeuralNetworks.jl/blob/master/src/msgpass.jl
is framework agnostic, Lux's layers can be implemented in terms of those.
I am not planning to port GNN.jl to Lux.jl, but yes, it would be nice to have a gnn library based on Lux, hopefully sharing as much codebase as possible with existing packages

Reply:
@MilkshakeForReal I'm not sure what you mean, can you open an issue in GNN.jl so that we can discuss there possible design issues?

Reply:
Sure 😃